File: freighter/go/errors_test.go
// Copyright 2025 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package freighter_test

import (
	"context"

	. "github.com/onsi/ginkgo/v2"
	. "github.com/onsi/gomega"
	"github.com/synnaxlabs/freighter"
	"github.com/synnaxlabs/x/errors"
)

var _ = Describe("Errors", func() {
	var ctx context.Context
	BeforeEach(func() {
		ctx = context.Background()
	})
	Describe("Encoding", func() {
		It("Should encode EOF", func() {
			pld := errors.Encode(ctx, freighter.EOF, false)
			Expect(pld.Type).To(Equal("freighter.eof"))
			Expect(pld.Data).To(ContainSubstring("EOF"))
		})
		It("Should encode ErrStreamClosed", func() {
			pld := errors.Encode(ctx, freighter.ErrStreamClosed, false)
			Expect(pld.Type).To(Equal("freighter.stream_closed"))
			Expect(pld.Data).To(ContainSubstring("stream closed"))
		})
		It("Shouldn't encode unknown errors", func() {
			pld := errors.Encode(ctx, errors.New("unknown error"), false)
			Expect(pld.Type).To(Equal(errors.TypeUnknown))
			Expect(pld.Data).To(ContainSubstring("unknown error"))
		})
	})
	Describe("Decoding", func() {
		It("Should decode EOF", func() {
			pld := errors.Payload{Type: "freighter.eof"}
			err := errors.Decode(ctx, pld)
			Expect(err).To(MatchError(freighter.EOF))
			Expect(err.Error()).To(Equal("EOF"))
		})
		It("Should decode ErrStreamClosed", func() {
			pld := errors.Payload{Type: "freighter.stream_closed"}
			err := errors.Decode(ctx, pld)
			Expect(err).To(MatchError(freighter.ErrStreamClosed))
			Expect(err.Error()).To(ContainSubstring("stream closed"))
		})
		It("should decode unknown freighter error", func() {
			pld := errors.Payload{Type: "freighter.unknown", Data: "unknown error"}
			err := errors.Decode(ctx, pld)
			Expect(err.Error()).To(Equal("unknown error"))
		})
		It("shouldn't decode unknown error", func() {
			pld := errors.Payload{Type: "unknown", Data: "unknown error"}
			err := errors.Decode(ctx, pld)
			Expect(err.Error()).To(Equal("unknown error"))
		})
	})
})

File: driver/opc/errors/errors_test.cpp
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

#include "gtest/gtest.h"
#include "open62541/types.h"

#include "x/cpp/test/test.h"

#include "driver/opc/errors/errors.h"

namespace driver::opc::errors {
/// @brief it should return nil error for good status code.
TEST(ErrorTest, testParseErrorGood) {
    ASSERT_NIL(parse(UA_STATUSCODE_GOOD));
}

/// @brief it should map timeout status to unreachable error.
TEST(ErrorTest, testParseErrorConnectionTimeout) {
    ASSERT_OCCURRED_AS(parse(UA_STATUSCODE_BADTIMEOUT), UNREACHABLE);
}

/// @brief it should map not connected status to unreachable error.
TEST(ErrorTest, testParseErrorConnectionNotConnected) {
    ASSERT_OCCURRED_AS(parse(UA_STATUSCODE_BADNOTCONNECTED), UNREACHABLE);
}

/// @brief it should map secure channel closed status to unreachable error.
TEST(ErrorTest, testParseErrorConnectionSecureChannelClosed) {
    ASSERT_OCCURRED_AS(parse(UA_STATUSCODE_BADSECURECHANNELCLOSED), UNREACHABLE);
}

/// @brief it should map session invalid status to unreachable error.
TEST(ErrorTest, testParseErrorConnectionSessionInvalid) {
    ASSERT_OCCURRED_AS(parse(UA_STATUSCODE_BADSESSIONIDINVALID), UNREACHABLE);
}

/// @brief it should map session closed status to unreachable error.
TEST(ErrorTest, testParseErrorConnectionSessionClosed) {
    ASSERT_OCCURRED_AS(parse(UA_STATUSCODE_BADSESSIONCLOSED), UNREACHABLE);
}

/// @brief it should map session not activated status to unreachable error.
TEST(ErrorTest, testParseErrorConnectionSessionNotActivated) {
    ASSERT_OCCURRED_AS(parse(UA_STATUSCODE_BADSESSIONNOTACTIVATED), UNREACHABLE);
}

/// @brief it should map connection rejected status to unreachable error.
TEST(ErrorTest, testParseErrorConnectionRejected) {
    ASSERT_OCCURRED_AS(parse(UA_STATUSCODE_BADCONNECTIONREJECTED), UNREACHABLE);
}

/// @brief it should map disconnect status to unreachable error.
TEST(ErrorTest, testParseErrorDisconnect) {
    ASSERT_OCCURRED_AS(parse(UA_STATUSCODE_BADDISCONNECT), UNREACHABLE);
}

/// @brief it should map connection closed status to unreachable error.
TEST(ErrorTest, testParseErrorConnectionClosed) {
    ASSERT_OCCURRED_AS(parse(UA_STATUSCODE_BADCONNECTIONCLOSED), UNREACHABLE);
}

/// @brief it should map out of range status to critical error.
TEST(ErrorTest, testParseErrorNonConnectionError) {
    ASSERT_OCCURRED_AS(parse(UA_STATUSCODE_BADOUTOFRANGE), CRITICAL);
}

/// @brief it should map unknown node id status to critical error.
TEST(ErrorTest, testParseErrorInvalidNodeId) {
    ASSERT_OCCURRED_AS(parse(UA_STATUSCODE_BADNODEIDUNKNOWN), CRITICAL);
}

/// @brief it should map user access denied status to critical error.
TEST(ErrorTest, testParseErrorAccessDenied) {
    ASSERT_OCCURRED_AS(parse(UA_STATUSCODE_BADUSERACCESSDENIED), CRITICAL);
}
}
